jetty-core/jetty-http3/jetty-http3-client-transport/src/main/java/org/eclipse/jetty/http3/client/transport/internal/HttpReceiverOverHTTP3.java

```diff
@@ -20,44 +20,73 @@
 import org.eclipse.jetty.client.HttpResponse;
 import org.eclipse.jetty.http.HttpField;
 import org.eclipse.jetty.http.HttpFields;
-import org.eclipse.jetty.http.HttpStatus;
 import org.eclipse.jetty.http.MetaData;
 import org.eclipse.jetty.http3.api.Stream;
 import org.eclipse.jetty.http3.frames.HeadersFrame;
-import org.eclipse.jetty.http3.internal.HTTP3ErrorCode;
-import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.io.Content;
 import org.eclipse.jetty.util.Promise;
-import org.eclipse.jetty.util.thread.Invocable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 public class HttpReceiverOverHTTP3 extends HttpReceiver implements Stream.Client.Listener
 {
     private static final Logger LOG = LoggerFactory.getLogger(HttpReceiverOverHTTP3.class);
 
+    private boolean responseSucceeded;
+
     protected HttpReceiverOverHTTP3(HttpChannelOverHTTP3 channel)
     {
         super(channel);
     }
 
     @Override
-    protected HttpChannelOverHTTP3 getHttpChannel()
+    protected void onInterim()
     {
-        return (HttpChannelOverHTTP3)super.getHttpChannel();
     }
 
     @Override
-    protected void receive()
+    protected void reset()
     {
-        // Called when the application resumes demand of content.
+        super.reset();
+        responseSucceeded = false;
+    }
+
+    @Override
+    public Content.Chunk read(boolean fillInterestIfNeeded)
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("Reading, fillInterestIfNeeded={} in {}", fillInterestIfNeeded, this);
+        Stream stream = getHttpChannel().getStream();
+        Stream.Data data = stream.readData();
         if (LOG.isDebugEnabled())
-            LOG.debug("resuming response processing on {}", this);
+            LOG.debug("Read stream data {} in {}", data, this);
+        if (data == null)
+        {
+            if (fillInterestIfNeeded)
+                stream.demand();
+            return null;
+        }
+        ByteBuffer byteBuffer = data.getByteBuffer();
+        boolean last = !byteBuffer.hasRemaining() && data.isLast();
+        if (last && !responseSucceeded)
+        {
+            responseSucceeded = true;
```

Comment:
Is field `responseSucceeded` necessary? We should be protected by reads after EOF by the base class.

Reply:
No, it's not needed. Removed it.